1.15.x recipe book categories stuff

nils-prommersberger opened the pull request:
The Modder is now able to create his own RecipeBookCategories.
He can add them to MinecraftForge during the FMLClientSetupEvent with RecipeBookCategories#addAdditionalCategory. There he have also to add the condition. The Modder can also use that method with condition to add his own Recipes to Vanillas RecipeBookCategories when they do not fit in automatically. 
Modder is also able to use vanillas recipe unlock stuff.


PS: 
I do not know where to place my example mod, so I placed it in the test_module

nils-prommersberger committed:
Merge pull request #1 from MinecraftForge/1.15.x

update

Commit:
just all my sutff

Commit:
oops

Commit:
just a little smaller and some not needed stuff

Commit:
optimise and bug fixing

nils-prommersberger committed:
Update RecipeBookCategories.java.patch

allow one condition on more than one category because vanilla does the same thing with all sub-categories and a search category

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=6669) <br/>All committers have signed the CLA.

diesieben07 commented on patches/minecraft/net/minecraft/client/util/ClientRecipeBook.java.patch:
```diff
@@ -0,0 +1,74 @@
+ import java.util.Collections;
+ import java.util.List;
+ import java.util.Map;
++import java.util.function.Function;
```
No imports in patch files.

diesieben07 commented on patches/minecraft/net/minecraft/client/util/ClientRecipeBook.java.patch:
```diff
@@ -0,0 +1,74 @@
+ import java.util.List;
+ import java.util.Map;
++import java.util.function.Function;
++
```
don't add newlines

diesieben07 commented on patches/minecraft/net/minecraft/client/util/RecipeBookCategories.java.patch:
```diff
@@ -0,0 +1,79 @@
+-   SMOKER_FOOD(new ItemStack(Items.field_151147_al)),
+-   STONECUTTER(new ItemStack(Items.field_221782_da)),
+-   CAMPFIRE(new ItemStack(Items.field_151147_al));
++public class RecipeBookCategories {
```
No, just no.
Use `IExtensibleEnum`.

gigaherz reviewed (changes requested):
Besides the already mentioned issues in the code, this is missing a critical feature if mods are allowed to add new categories: an improved GUI which is capable of handling more than 5 categories without overflowing the vertical space.

For reference, I already implemented an equivalent feature back in January, which I didn't push because of the lack of an improved GUI. 

It also shows how to make use of the extensible enum system: https://github.com/MinecraftForge/MinecraftForge/compare/1.15.x...gigaherz:custom-machine-recipebook

Commit:
remove all extra-imports
optimise!!!

nils-prommersberger commented:
@gigaherz I just want the ability for modders to add their own RecipeBookCategories. To implement these in vanilla guis or their own is left to the modders.

In the ClientRecipeBook class I changed the name of func_216767_a but I don't believe that is the right way. What would be the right way if it is not?

So I implemented IExtensibleEnum (I mean it's cool, thanks) - but how to get access to the factory in my test mod?

tterrag1098 commented:
"compatibility is left to modders" is not good enough for forge. If this is going to be officially supported it needs to be compatible out of the box, and that includes handling more than 5 categories.

nils-prommersberger committed:
Update ClientRecipeBook.java.patch

small failure

nils-prommersberger commented:
I think I said it wrong. 

Until now, when unlocking a recipe in Minecraft via "minecraft:unlock_recipe", the recipe is always displayed in the crafting table. First I just wanted to open the possibility to use "minecraft:unlock_recipe" and the RecipeBook, not the graphics. But if this is still not enough for you after this explanation, I will try to get an extended implementation.

Commit:
ability of search category
IExtensibleEnum implementation

Commit:
Merge remote-tracking branch 'origin/1.15.x-RecipeBookCategoriesStuff' into 1.15.x-RecipeBookCategoriesStuff

Commit:
bug fixing

Commit:
update

Commit:
update

Commit:
update

nils-prommersberger commented:
ok it's  not ready to merge. But I would like to know if everything is good so far or if I should change something.

pupnewfster commented on patches/minecraft/net/minecraft/client/gui/recipebook/RecipeBookGui.java.patch:
```diff
@@ -9,7 +39,113 @@
+@@ -189,24 +202,68 @@
+    }
+ 
+    private void func_193949_f() {
```
Rather than making such a large/"messy" patch. It may make more sense to make the new logic in its own method at the bottom of the RecipeBookGui file, and then just patch calls to this method to instead call that new method.

niprow commented:
ups sorry
I'll take care of it

Commit:
update

Commit:
ability of setting the categories of a type

Commit:
category do not have to be a subcategory of the first search category of the block

nils-prommersberger committed:
Merge pull request #2 from MinecraftForge/1.15.x

update to newest brunch

Commit:
Merge remote-tracking branch 'origin/1.15.x' into 1.15.x-RecipeBookCategoriesStuff

# Conflicts:
#	src/test/resources/META-INF/mods.toml

Commit:
optimize test

ichttt commented:
Screenshots would be nice.
Also, replacing the GUI isn't that good of a solution because you have to copy the vanilla file and redistribute it with minor changes, and it breaks resource packs. Maybe you can get away with using and in-code rotating the existing arrow textures
